renamed variables in lq_cut, changed expression handling

DaGeibl opened the pull request:
Some changes for the lq_cut

- renamed some variables in lq_cut, and renamed the drift time correction to function `lq_ctc_correction`
- lq_cut now requires `lq_e_corr_expression `and `dt_eff_expression` to construct prop func
- added the option to use higher order corrections for the ctc fit
- use`cut_single_peak` insted of own pre-cutting method
- adjusted recipe to changes

theHenks commented:
All tests are failing. I think you need to rename things in there too

theHenks commented on ext/LegendSpecFitsRecipesBaseExt.jl:
```diff
@@ -1390,7 +1390,7 @@ end
         xlabel := "Energy"
         ylabel := "LQ (A.U.)"
-        nbins := (0:6:3000, -2.0:0.012:4.0)
+        nbins := (0:6:3000, -0.1:0.0006:0.2)
```
Can you do the binning dynamically at least for the `LQ`? I encountered already several issues with extreme RAM usage during the dataflow when algorhtihms are binning "non-normal" data into extremely fine bins which allocates a lot of memory?

fhagemann replied:
Well, it looks like here it is always the same "fineness" of binning and the only change is dividing the LQ values by 20 (why is that?)

theHenks commented on src/lqcut.jl:
```diff
@@ -1,54 +1,46 @@
+    lq_norm::Vector{<:AbstractFloat}, dt_eff::Vector{<:Unitful.RealOrRealQuantity}, e_cal::Vector{<:Unitful.Energy{<:Real}}, DEP_µ::Unitful.AbstractQuantity, DEP_σ::Unitful.AbstractQuantity; 
+    ctc_dep_edgesigma::Float64=3.0 , ctc_driftime_cutoff_method::Symbol=:percentile, lq_outlier_sigma::Float64 = 2.0, drift_time_outlier_sigma::Float64 = 2.0, prehist_sigma::Float64=2.5, lq_e_corr_expression::Union{String,Symbol}="(lq / e)", dt_eff_expression::Union{String,Symbol}="(qdrift / e)" ,ctc_dt_eff_low_quantile::Float64=0.15, ctc_dt_eff_high_quantile::Float64=0.95, pol_fit_order::Int=1) 
 
 Perform the drift time correction on the LQ data using the DEP peak. The function cuts outliers in lq and drift time, then performs a linear fit on the remaining data. The data is Corrected by subtracting the linear fit from the lq data.
```
Could you add a description of the `kwargs`?

fhagemann replied:
I agree

theHenks commented on src/lqcut.jl:
```diff
@@ -1,54 +1,46 @@
-    lq_prestats = estimate_single_peak_stats(lq_prehist)
-    lq_start = lq_prestats.peak_pos - prehist_sigma * lq_prestats.peak_sigma
-    lq_stop = lq_prestats.peak_pos + prehist_sigma * lq_prestats.peak_sigma
+    lq_precut = cut_single_peak(lq_DEP, minimum(lq_DEP), maximum(lq_DEP))
```
The `cut_single_peak` function offers a couple `kwargs` which can be tuned. You might want to have a look into that if you want to parse them through here. 
Also, maybe you can find preset parameters for the big windowing rather than just `minimum(lq_DEP)` and `maximum(lq_DEP)`

theHenks commented on src/lqcut.jl:
```diff
@@ -1,54 +1,46 @@
 
     # truncated gaussian fit
-    lq_result, lq_report = fit_single_trunc_gauss(lq_DEP, (low=lq_start, high=lq_stop, max=NaN))
+    lq_result, lq_report = fit_single_trunc_gauss(lq_DEP, lq_precut)
```
You can set `uncertainty=false` here to speed up the fit if you don't need the uncertainties.

theHenks commented on src/lqcut.jl:
```diff
@@ -65,10 +57,10 @@ function lq_drift_time_correction(
+        t_upper = µ_t + drift_time_outlier_sigma * σ_t
 
-    elseif mode == :double_gaussian # can be used for detectors with double peaks; not optimized yet
+    elseif ctc_driftime_cutoff_method == :double_gaussian # can be used for detectors with double peaks; not optimized yet
```
Maybe also consider here setting `uncertainty=false` if you are only interested in the Best fit values.

theHenks commented on src/lqcut.jl:
```diff
@@ -98,13 +90,13 @@ function lq_drift_time_correction(
+    #polynomial fit
+    result_µ, report_µ = chi2fit(pol_fit_order, t_cut, lq_cut; uncertainty=true)
+    par = mvalue(result_µ.par)
+    drift_time_func(x) =  sum((mvalue(par[i])) * x^(i-1) for i in eachindex(par))
```
Is this supposed to just be the Polynomial? Because for that there is the `PolCalFunc` which you can use. In addition, I think the `chi2` report contains the function too.

theHenks commented on src/lqcut.jl:
```diff
@@ -1,54 +1,46 @@
-    lq_norm::Vector{Float64}, dt_eff::Vector{<:Unitful.RealOrRealQuantity}, e_cal::Vector{<:Unitful.Energy{<:Real}}, DEP_µ::Unitful.AbstractQuantity, DEP_σ::Unitful.AbstractQuantity; 
-    DEP_edgesigma::Float64=3.0 , mode::Symbol=:percentile, drift_cutoff_sigma::Float64 = 2.0, prehist_sigma::Float64=2.5, e_expression::Union{String,Symbol}="e", dt_eff_low_quantile::Float64=0.15, dt_eff_high_quantile::Float64=0.95)
+    lq_norm::Vector{<:AbstractFloat}, dt_eff::Vector{<:Unitful.RealOrRealQuantity}, e_cal::Vector{<:Unitful.Energy{<:Real}}, DEP_µ::Unitful.AbstractQuantity, DEP_σ::Unitful.AbstractQuantity; 
+    ctc_dep_edgesigma::Float64=3.0 , ctc_driftime_cutoff_method::Symbol=:percentile, lq_outlier_sigma::Float64 = 2.0, drift_time_outlier_sigma::Float64 = 2.0, prehist_sigma::Float64=2.5, lq_e_corr_expression::Union{String,Symbol}="(lq / e)", dt_eff_expression::Union{String,Symbol}="(qdrift / e)" ,ctc_dt_eff_low_quantile::Float64=0.15, ctc_dt_eff_high_quantile::Float64=0.95, pol_fit_order::Int=1) 
```
Maybe we can stay with `qdrift` instead of `dt_eff` to avoid confusion with the other routines.

fhagemann replied:
A docstring should usually have the name of the function in it (in this case `lq_ctc_correction`).
Please add that to be consistent with other docstrings. Also have a look at #80

DaGeibl replied:
Ok, but then i would just use dt_eff to not confuse it with qdrift

theHenks commented on src/lqcut.jl:
```diff
@@ -168,8 +159,10 @@ function LQ_cut(
+    #simplecuts benutzen
 
     # Fit the sideband subtracted histogram
     fit_result, fit_report = fit_binned_trunc_gauss(hist_corrected, (low=lq_start, high=lq_stop, max=NaN))
```
I think, uncertainties here would be nice.

theHenks replied:
I would also add more stuff into the result, in particular about this fit. Since otherwise, you cannot debug anything later.

DaGeibl replied:
The `fit_binned_trunc_gauss` returns values with uncertainties or do you mean something else?

theHenks replied:
Yes, but set the `uncertainty=true` kwarg explicitly.

DaGeibl replied:
The fit is saved in the report. In gerneral what should be inside the result next to final values, like the cut, or the property function to get the classifier?

theHenks replied:
Whatever you put into the result, will be later saved on disk and can be accessed for each run. This is then the only chance to do monitoring etc. So you should think about what you want to plot later and then fill the `result` accordingly.

theHenks reviewed (changes requested):
I have mainly some comments about the uncertainty handling, the returns and some of the naming.

fhagemann commented on src/lqcut.jl:
```diff
@@ -1,54 +1,46 @@
-
-    # get energy units to remve later
-    e_unit = unit(first(e_cal))
+    ctc_dep_edgesigma::Float64=3.0 , ctc_driftime_cutoff_method::Symbol=:percentile, lq_outlier_sigma::Float64 = 2.0, drift_time_outlier_sigma::Float64 = 2.0, prehist_sigma::Float64=2.5, lq_e_corr_expression::Union{String,Symbol}="(lq / e)", dt_eff_expression::Union{String,Symbol}="(qdrift / e)" ,ctc_dt_eff_low_quantile::Float64=0.15, ctc_dt_eff_high_quantile::Float64=0.95, pol_fit_order::Int=1) 
```
why `ctc`?

DaGeibl replied:
I copied the variable names from florians preliminary config (https://github.com/legend-exp/legend-jldataflow-config/blob/main/psd/p0%25_r0%25%25_psd.json). 
In general what pygama calls drift time correction we call carge trapping correction right?(e.g. for the A/E) I can also revert it and call everything drift time correction, dependant on what you find more understandable/ or consistent.

theHenks replied:
This is just a first proposal based on your old version. Please update the `jldataprod/config` accordingly with your new naming putting all your metadata pars in there and link the PR here.

fhagemann commented on src/lqcut.jl:
```diff
@@ -127,23 +118,23 @@ function lq_drift_time_correction(
 
 """
-    DEP_µ::Unitful.Energy, DEP_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{Float64}; cut_sigma::Float64=3.0, truncation_sigma::Float64=2.0)
+    DEP_µ::Unitful.Energy, DEP_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{<:AbstractFloat}; cut_sigma::Float64=3.0, dep_sideband_sigma::Float64=4.5, cut_truncation_sigma::Float64=2.0)
```
Again here: the docstring is missing the name of the function: `lq_cut` and the description of the arguments (see #80)

fhagemann commented on src/lqcut.jl:
```diff
@@ -127,23 +118,23 @@ function lq_drift_time_correction(
-    DEP_µ::Unitful.Energy, DEP_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{Float64}; cut_sigma::Float64=3.0, truncation_sigma::Float64=2.0)
+    DEP_µ::Unitful.Energy, DEP_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{<:AbstractFloat}; cut_sigma::Float64=3.0, dep_sideband_sigma::Float64=4.5, cut_truncation_sigma::Float64=2.0)
 
 Evaluates the cutoff value for the LQ cut. The function performs a binned gaussian fit on the sidebandsubtracted LQ histogram and evaluates the cutoff value difined at 3σ of the fit.
 # Returns
     * `result`: NamedTuple of the cutoff value
     * `report`: NamedTuple of the fit result, fit report and temporary histograms
 """
-function LQ_cut(
-    DEP_µ::Unitful.Energy, DEP_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{<:AbstractFloat}; cut_sigma::Float64=3.0, truncation_sigma::Float64=2.0)
+function lq_cut(
+    DEP_µ::Unitful.Energy, DEP_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{<:AbstractFloat}; cut_sigma::Float64=3.0, dep_sideband_sigma::Float64=4.5, cut_truncation_sigma::Float64=2.0)
 
     # Define sidebands
-    lq_DEP = lq_classifier[DEP_µ - 4.5 * DEP_σ .< e_cal .< DEP_µ + 4.5 * DEP_σ]
-    lq_sb1 = lq_classifier[DEP_µ -  2 * 4.5 * DEP_σ .< e_cal .< DEP_µ - 4.5 * DEP_σ]
-    lq_sb2 = lq_classifier[DEP_µ + 4.5 * DEP_σ .< e_cal .< DEP_µ + 2 * 4.5 * DEP_σ]
+    lq_DEP = lq_classifier[DEP_µ - dep_sideband_sigma * DEP_σ .< e_cal .< DEP_µ + dep_sideband_sigma * DEP_σ]
+    lq_sb1 = lq_classifier[DEP_µ -  2 * dep_sideband_sigma * DEP_σ .< e_cal .< DEP_µ - dep_sideband_sigma * DEP_σ]
+    lq_sb2 = lq_classifier[DEP_µ + dep_sideband_sigma * DEP_σ .< e_cal .< DEP_µ + 2 * dep_sideband_sigma * DEP_σ]
```
Could you somehow make the casing (uppercase/lowercase) of `DEP`/`dep` consistent?

DaGeibl replied:
Would you suggest in variables lower case for all variables, except the "peak_names"? Because there they usually appear uppercase

theHenks replied:
Whatever choice you take, it should be consistent.

DaGeibl replied:
ok then I will write dep in variables lowercase, because all variables are typically lowercase,but when talking about it e.g in docstrings have the abbreviation uppercase

theHenks commented on ext/LegendSpecFitsRecipesBaseExt.jl:
```diff
@@ -1275,7 +1275,7 @@ end
 # recipe for the lq_drift_time_correction report
 
-@recipe function f(report::NamedTuple{(:lq_prehist, :lq_report, :drift_prehist, :drift_report, :lq_box, :drift_time_func, :DEP_left, :DEP_right)}, e_cal, dt_eff, lq_e_corr, plot_type::Symbol)
+@recipe function f(report::NamedTuple{(:lq_report, :drift_report, :lq_box, :drift_time_func, :DEP_left, :DEP_right)}, e_cal, dt_eff, lq_e_corr, plot_type::Symbol)
```
What is the `plot_type` doing? Can this be a `kwarg`?

DaGeibl replied:
`plot_type` specifies which one the different plots should be plottet. 
Should i implement this by picking one them as a default plot?

theHenks commented on src/lqcut.jl:
```diff
@@ -88,23 +92,23 @@ function lq_drift_time_correction(
         t_upper = maximum(x_at_threshold)
     else
-        error("Mode $mode not supported")
+        error("Driftime cutoff method $ctc_driftime_cutoff_method not supported")
```
If you want that the program exits here, you have to throw an `ArgumentError`

theHenks commented on src/lqcut.jl:
```diff
@@ -114,71 +118,86 @@ function lq_drift_time_correction(
     )
```
There are still things in the `result` missing in my opinion. What is with the GOF, some debug parameters from the correction, the fit etc? 
I guess it is interesting to look later on at least at some variables and their behavior over time. Right now, nothing apart from the string will be saved.

theHenks commented on src/lqcut.jl:
```diff
@@ -114,71 +118,86 @@ function lq_drift_time_correction(
     result = (
         cut = cut_3σ,
+        fit_µ = fit_result.μ,
+        fit_σ = fit_result.σ,
```
Same comment here as above, at least information about the fit is necessary.

theHenks commented on src/singlefit.jl:
```diff
@@ -473,6 +473,7 @@ function fit_binned_trunc_gauss(h_nocut::Histogram, cuts::NamedTuple{(:low, :hig
     report = (
             f_fit = x -> Base.Fix2(f_fit, v_ml)(x) * bin_width,
             h = h_nocut,
+            h_cut = h,
```
There is a recipe for this, did you change that too so that it still produces a plot?

DaGeibl replied:
I updated the recipe

theHenks reviewed (changes requested):
The `results` still need a fix.

fhagemann commented on src/aoe_cut.jl:
```diff
@@ -199,25 +200,24 @@ Get the survival fraction of a continuum after a AoE cut value `aoe_cut` for a g
     # get energy around center
-    aoe = aoe[center - window .< e .< center + window]
-    e = e[center - window .< e .< center + window]
+    e_wdw = e[center - window .< e .< center + window]
     # get bin width
-    bin_width = get_friedman_diaconis_bin_width(e)
+    bin_width = get_friedman_diaconis_bin_width(e_wdw)
     # get number of events in window before cut
-    n_before = length(e)
+    n_before = length(e_wdw)
     # get energy after cuts
     e_survived, e_cut = if !inverted_mode
         #normal aoe version
         e[aoe_cut .< aoe .< sigma_high_sided], e[aoe .<= aoe_cut .|| aoe .>= sigma_high_sided]
     else
         #lq version
-        e[aoe .< aoe_cut .|| aoe .> sigma_high_sided], e[aoe_cut .<= aoe .<= sigma_high_sided]
+        e[aoe .< aoe_cut .&& .!aoe_lowcut_vec], e[aoe_cut .<= aoe .&& .!aoe_lowcut_vec]
     end
-    n_after = length(e_survived)
+    n_after = length(e_survived[center - window .< e_survived .< center + window])
```
How about something like
```
e_wdw_idx = findall(center - window .< e .< center + window)
```
and then index arrays with `e_wdw_idx` where needed?

DaGeibl commented:
I now also changed in function `aoe `to `psd_parameter `and `aoe_cut `to `psd_cut `to make clear it is also used for lq(and maybe future cuts) and not only aoe. if you want a shorter name than psd_parameter like `psd_par `or `psd_param `let me know then i change it

DaGeibl commented:
Compared to the last pull rewuest i removed the Union{Bitvector, nothing} type restriction and changed it to just Bitvector. In this version the `get_peaks_survival_fractions` needs two different versions, one with and one without Bitvector. I am not sure if this can be solved more elegeant as i did.

DaGeibl commented:
@theHenks @fhagemann This pull request is finished from my side, i would be happy if you could review it

fhagemann commented on src/aoe_cut.jl:
```diff
@@ -242,5 +356,32 @@ function get_continuum_survival_fraction(aoe::Vector{<:Unitful.RealOrRealQuantit
     return result, report
 end
 
+# Overloaded function for psd_cut only
+function get_continuum_survival_fraction(psd_parameter::Vector{<:Unitful.RealOrRealQuantity}, e::Vector{<:T}, center::T, window::T, psd_cut::Unitful.RealOrRealQuantity; 
+    inverted_mode::Bool=false, sigma_high_sided::Unitful.RealOrRealQuantity=Inf*unit(first(psd_parameter))) where T<:Unitful.Energy{<:Real}
+
+    cut_vector = if !inverted_mode
+        psd_cut .< psd_parameter .< sigma_high_sided
+    else
+        psd_parameter .< psd_cut
+    end
+    get_continuum_survival_fraction(cut_vector, psd_parameter, e, center, window, psd_cut; 
+                                    sigma_high_sided=sigma_high_sided)
+end
+
+# Overloaded function for psd_cut and existing cut_vector
+function get_continuum_survival_fraction(psd_parameter::Vector{<:Unitful.RealOrRealQuantity}, e::Vector{<:T}, center::T, window::T, psd_cut::Unitful.RealOrRealQuantity, cut_vector::BitVector; 
+    inverted_mode::Bool=false, sigma_high_sided::Unitful.RealOrRealQuantity=Inf*unit(first(psd_parameter))) where T<:Unitful.Energy{<:Real}
+
+    combined_cut_vector = if !inverted_mode
+        psd_cut .< psd_parameter .< sigma_high_sided .&& cut_vector
+    else
+        psd_parameter .< psd_cut .&& cut_vector
+    end
+
+    get_continuum_survival_fraction(combined_cut_vector, psd_parameter, e, center, window, psd_cut; 
+                                    sigma_high_sided=sigma_high_sided)
+end
```
Again, `cut_vector` could obtain a default value of `trues(length(psd_parameter))` to avoid having two functions.

fhagemann replied:
Same things as discussed in https://github.com/legend-exp/LegendSpecFits.jl/pull/107#discussion_r1956241337

fhagemann commented on src/aoe_cut.jl:
```diff
@@ -88,14 +88,59 @@ export get_low_aoe_cut
 - `report`: Dict of reports for each peak
 """
-function get_peaks_survival_fractions(aoe::Vector{<:Unitful.RealOrRealQuantity}, e::Vector{<:T}, peaks::Vector{<:T}, peak_names::Vector{Symbol}, windows::Vector{<:Tuple{T, T}}, aoe_cut::Unitful.RealOrRealQuantity,; uncertainty::Bool=true, inverted_mode::Bool=false, bin_width_window::T=2.0u"keV", sigma_high_sided::Unitful.RealOrRealQuantity=Inf*unit(first(aoe)), fit_funcs::Vector{Symbol}=fill(:gam ...
+function get_peaks_survival_fractions(psd_parameter::Vector{<:Unitful.RealOrRealQuantity}, e::Vector{<:T}, peaks::Vector{<:T}, peak_names::Vector{Symbol}, windows::Vector{<:Tuple{T, T}}, psd_cut::Unitful.RealOrRealQuantity, cut_vector::BitVector,; 
+    uncertainty::Bool=true, inverted_mode::Bool=false, bin_width_window::T=2.0u"keV", sigma_high_sided::Unitful.RealOrRealQuantity=Inf*unit(first(psd_parameter)), fit_funcs::Vector{Symbol}=fill(:gamma_def, length(peaks))) where T<:Unitful.Energy{<:Real}
+    @assert length(peaks) == length(peak_names) == length(windows) "Length of peaks, peak_names and windows must be equal"
+    # create return and result vectors
+    v_result = Vector{NamedTuple}(undef, length(peak_names))
+    v_report = Vector{NamedTuple}(undef, length(peak_names))
+    
+    # iterate throuh all peaks
+    Threads.@threads for i in eachindex(peak_names)
+        # extract peak, name and window
+        peak, name, window, fit_func = peaks[i], peak_names[i], windows[i], fit_funcs[i]
+        # fit peak
+        result_peak, report_peak = get_peak_survival_fraction(psd_parameter, e, peak, collect(window), psd_cut, cut_vector; uncertainty=uncertainty, inverted_mode=inverted_mode, bin_width_window=bin_width_window, sigma_high_sided=sigma_high_sided, fit_func=fit_func)
+        # save results
+        v_result[i] = result_peak
+        v_report[i] = report_peak
+    end
+
+    # create result and report dict
+    result = Dict{Symbol, NamedTuple}(peak_names .=> v_result)
+    report = Dict{Symbol, NamedTuple}(peak_names .=> v_report)
+    
+    return result, report
+end
```
Just as noted below, `cut_vector` could for example receive a default value of `trues(length(psd_parameter))` to avoid having two methods that basically do the same thing (to avoid code duplication).

DaGeibl replied:
Do I then even need two functions at all? can't i just have one function that takes psd_parameter, and an optional cut vector?

fhagemann replied:
Nope, you would just need the one having the new implementation of `cut_vector`, with a default value if no value for `cut_vector` is passed.

DaGeibl replied:
Ok to be clear instead of now 3 and I could go to 2 or 1? Because say i could just use 1 with the optional cut vector.

fhagemann replied:
You can also take the code from `get_peak_survival_fraction(cut_vector, ...)` and put it back into the original
`get_peak_survival_fraction(..., cut_vector = trues(length(psd_parameter)))` method.

DaGeibl replied:
What I was saying is, that i could could just merge all 3 `get_peak_survival_fraction` functions into a single one that has a `cut_vector = trues(length(psd_parameter)))` as argument. To not have the `_get_peak_survival_fraction function`, only a single one that does everything

fhagemann replied:
That sounds good to me 👍

fhagemann replied:
Let me know once this is pushed so I can review this PR in the final round.

fhagemann commented on src/lqcut.jl:
```diff
@@ -1,81 +1,88 @@
+    lq_dep_NaNfree = filter(isfinite, lq_dep)
+   
+    # precut lq data for fit
+    lq_precut = cut_single_peak(lq_dep_NaNfree, minimum(lq_dep_NaNfree), quantile(lq_dep_NaNfree, 0.99); relative_cut=ctc_lq_precut_relative_cut)
```
Should the `0.99` become a keyword argument, so that it can be changed by the user if wanted?

DaGeibl replied:
i mean i could, but it's only there to remove extreme outliers so I don't i would ever modify that number

fhagemann commented on src/lqcut.jl:
```diff
@@ -88,97 +95,111 @@ function lq_drift_time_correction(
     end
 
     #store cutoff values in box to return later    
     box = (lq_lower = lq_lower, lq_upper = lq_upper, t_lower = t_lower, t_upper = t_upper)
```
No need to change that here, but if you create a named tuple that has the field names identical to the variables that you're setting, you could write this in a more compact way:
```julia
box = (; lq_lower, lq_upper, t_lower, t_upper)
```

fhagemann commented on src/lqcut.jl:
```diff
@@ -88,97 +95,111 @@ function lq_drift_time_correction(
-function LQ_cut(
-    DEP_µ::Unitful.Energy, DEP_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{<:AbstractFloat}; cut_sigma::Float64=3.0, truncation_sigma::Float64=2.0)
+function lq_cut(
+    dep_µ::Unitful.Energy, dep_σ::Unitful.Energy, e_cal::Vector{<:Unitful.Energy}, lq_classifier::Vector{<:AbstractFloat}; cut_sigma::Float64=3.0, dep_sideband_sigma::Float64=4.5, cut_truncation_sigma::Float64=3.5)
```
A quick comment why you changed the default value of the `cut_truncation_sigma` from `2.0` to `3.5`?

fhagemann commented on src/lqcut.jl:
```diff
@@ -88,97 +95,111 @@ function lq_drift_time_correction(
     
     # Create a named tuple of histograms for crosschecks
-    temp_hists = (hist_DEP = hist_DEP, hist_sb1 = hist_sb1, hist_sb2 = hist_sb2, hist_subtracted = hist_subtracted, hist_corrected = hist_corrected)
+    temp_hists = (hist_dep = hist_dep, hist_sb1 = hist_sb1, hist_sb2 = hist_sb2, hist_subtracted = hist_subtracted, hist_corrected = hist_corrected)
```
This notation could also be compactified

fhagemann commented on src/lqcut.jl:
```diff
@@ -88,97 +95,111 @@ function lq_drift_time_correction(
 
     # Fit the sideband subtracted histogram
-    fit_result, fit_report = fit_binned_trunc_gauss(hist_corrected, (low=lq_start, high=lq_stop, max=NaN))
+    fit_result, fit_report = fit_binned_trunc_gauss(hist_corrected, (low=lq_start, high=lq_stop, max=NaN), uncertainty=true)
```
make `uncertainty` a keyword argument of `lq_cut` and pass it here?

fhagemann commented on test/test_lq.jl:
```diff
@@ -29,12 +72,12 @@ using Distributions
 
-    # plot(report.temp_hists.hist_DEP, label="LQ SEP")
+    # plot(report.temp_hists.hist_dep, label="LQ SEP")
     # plot!(report.temp_hists.hist_sb1, label="LQ SB1")
     # plot!(report.temp_hists.hist_sb2, label="LQ SB2")
-    # plot!(report.temp_hists.hist_subtracted, label="DEP Subtracted")
+    # plot!(report.temp_hists.hist_subtracted, label="dep Subtracted")
     # plot(report.temp_hists.hist_corrected, label="original histogram")
     # plot!(report.fit_report.f_fit, label="Fit function")
```
Please remove from the tests what is not needed/used.

fhagemann reviewed (changes requested):
This PR features a lot of code duplication.
I commented where some functions can be combined to a single method by providing default values to the arguments.

Please also provide a plot using the new plot recipe, which makes it easier to review.

DaGeibl commented:
Here is an preview of the two plot recipes addedin this pull request.
Both plots are made using the report of LQ_cut. The first one is accesed by passing :sideband together with the report, lq_class and e_cal. It shows the sideband and the peak histogramms used for the sideband substraction. Here is an example:

![grafik](https://github.com/user-attachments/assets/1a643b7e-fe9b-4743-8d18-58f2039b5032)

The second plot is accesed by specifying :fit. 

![grafik](https://github.com/user-attachments/assets/c4c0ecc2-54ef-4292-9dd2-84f2e035b7bb)

DaGeibl commented on src/aoe_cut.jl:
```diff
@@ -88,14 +88,59 @@ export get_low_aoe_cut
+end
+get_peaks_survival_fractions(psd_parameter, e, peaks, peak_names, left_window_sizes::Vector{<:Unitful.Energy{<:Real}}, right_window_sizes::Vector{<:Unitful.Energy{<:Real}}, psd_cut, cut_vector; kwargs...) = get_peaks_survival_fractions(psd_parameter, e, peaks, peak_names, [(l,r) for (l,r) in zip(left_window_sizes, right_window_sizes)], psd_cut, cut_vector; kwargs...)
+
+function get_peaks_survival_fractions(psd_parameter::Vector{<:Unitful.RealOrRealQuantity}, e::Vector{<:T}, peaks::Vector{<:T}, peak_names::Vector{Symbol}, windows::Vector{<:Tuple{T, T}}, psd_cut::Unitful.RealOrRealQuantity,; 
+    uncertainty::Bool=true, inverted_mode::Bool=false, bin_width_window::T=2.0u"keV", sigma_high_sided::Unitful.RealOrRealQuantity=Inf*unit(first(psd_parameter)), fit_funcs::Vector{Symbol}=fill(:gamma_def, length(peaks))) where T<:Unitful.Energy{<:Real}
```
Yes excactly, to make clear, its not only aoe but all PSD cuts.

DaGeibl commented on src/lqcut.jl:
```diff
@@ -1,81 +1,88 @@
-    lq_start = lq_prestats.peak_pos - prehist_sigma * lq_prestats.peak_sigma
-    lq_stop = lq_prestats.peak_pos + prehist_sigma * lq_prestats.peak_sigma
+    #filter out NaN values
+    lq_dep_NaNfree = filter(isfinite, lq_dep)
```
I think the problem was that lq_cut and t_cut in line 105 and 106 are then not the same length

DaGeibl replied:
yeah that would work. But what exactly makes this better that the original way?

fhagemann replied:
you avoid creating another array `lq_dep_NaNfree` which would cause additional allocations and use of memory

fhagemann commented on src/aoe_cut.jl:
```diff
@@ -118,52 +163,62 @@ function get_peaks_survival_fractions(aoe::Vector{<:Unitful.RealOrRealQuantity},
-function get_peak_survival_fraction(aoe::Vector{<:Unitful.RealOrRealQuantity}, e::Vector{<:T}, peak::T, window::Vector{T}, aoe_cut::Unitful.RealOrRealQuantity,; 
-    uncertainty::Bool=true, inverted_mode::Bool=false, bin_width_window::T=2.0u"keV", sigma_high_sided::Unitful.RealOrRealQuantity=Inf*unit(first(aoe)), fit_func::Symbol=:gamma_def) where T<:Unitful.Energy{<:Real}  
-    # estimate bin width
+function get_peak_survival_fraction(cut_vector::BitVector, e::Vector{<:T}, peak::T, window::Vector{T}; 
+    uncertainty::Bool=true, bin_width_window::T=2.0u"keV", fit_func::Symbol=:gamma_def) where T<:Unitful.Energy{<:Real}  
```
You right now have three methods for `get_peak_survival_fraction`.
I would rename this one to some internal function name (appending `_` up front)
and for the docstring, you could do something like
```julia
"""
get_peaks_survival_fractions(psd_parameter::Vector{<:Unitful.RealOrRealQuantity}, e::Vector{<:T}, peaks::Vector{<:T}, peak_names::Vector{Symbol}, windows::Vector{<:Tuple{T, T}}, psd_cut::Unitful.RealOrRealQuantity[, cut_vector::BitVector]; 
    uncertainty::Bool=true, inverted_mode::Bool=false, bin_width_window::T=2.0u"keV", sigma_high_sided::Unitful.RealOrRealQuantity=Inf*unit(first(psd_parameter)), fit_funcs::Vector{Symbol}=fill(:gamma_def, length(peaks))) where T<:Unitful.Energy{<:Real}
 
Get the survival fraction of a peak after a PSD cut value `psd_cut` for a given `peak` and `window` size while performing a peak fit with fixed position.
# Arguments
- `psd_parameter`: Vector of PSD parameter values.
- `e`: Vector of energy values.
- `peaks`: Vector of peak positions in energy.
- `peak_names`: Vector of symbols representing peak names.
- `windows`: Vector of tuples specifying the lower and upper energy bounds for fitting each peak.
- `psd_cut`: The PSD threshold used for classification.
- `cut_vector` (optional): BitVector indicating which events pass the PSD cut. If no `cut_vector` is given, no additional cut is applied.
# Keyword Arguments
- `uncertainty`: Whether to compute uncertainties (default: `true`).
- `inverted_mode`: If `true`, inverts the PSD cut logic (default: `false`).
- `bin_width_window`: Bin width for histogramming within peak windows (default: `2 keV`).
- `sigma_high_sided`: Upper bound for peak fitting beyond `psd_cut` (default: `Inf`).
- `fit_funcs`: Vector specifying the peak fitting function for each peak (default: `:gamma_def`).
"""
```

fhagemann commented:
> The second plot is accesed by specifying :fit.
> 
> ![grafik](https://private-user-images.githubusercontent.com/131683460/413332810-c4c0ecc2-54ef-4292-9dd2-84f2e035b7bb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzk4MTcxMDgsIm5iZiI6MTczOTgxNjgwOCwicGF0aCI6Ii8xMzE2ODM0NjAvNDEzMzMyODEwLWM0YzBlY2MyLTU0ZWYtNDI5Mi05ZGQyLTg0ZjJlMDM1YjdiYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMjE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDIxN1QxODI2NDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYzlkYzVmNTQ3YWE1OWNhYzA5OGU2NmNiNDI1MTQxYjdkMWU5MzU5ZmExM2M2ZmMzN2QyODI1NDc0ZGU4NThmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.yyTJi3qH0F-EXPmkwEA2xJuNCpABLfJSv5PFIOi-gH0)

You could remove the xlabel from the upper plot and the title from the lower plot, and ideally make them touch.

DaGeibl commented:
@fhagemann i implemented the discussed changes to include a default cut vector of trues, to reduce the code duplication

fhagemann reviewed (approved):
@DaGeibl verified that the A/E output is not at all affected by the modifications in this PR, so this should be good to go from my side
